Implement DCW collections and allow full continent, country, and state names in -R -E

PaulWessel opened the pull request:
DCW _collections_ are named groups of countries (or states) that can be requested via their abbreviation or full name.  DCW 2.0.2 will ship with some standard collections but users can add their own in _~/.gmt/dcw-groups.txt which must have the format

```
# Any number of comments anywhere in the file
tag: abbreviation [full name]
XX,YY,ZZ,WW,...
```

where the XX,YY etc are valid DCW tags. Any number of tag/list sequences are allowed.  As a simple example, these two lines adds the collection of Scandinavian countries:

```
tag: SCA Scandinavia
NO,SE,DK
```

Once can then use the collection in one of two ways:

```
gmt coast -RSCA -Gred -B -pdf map
gmt coast -RScandinavia -Gred -B -pdf map
```

While deveopers will work on suitable collections to distribute as offical names, interested users can play with this branch and create collections as described.  I have placed an empty _dcw-collections.txt_ file on the data server so that we avoid errors when GMT tries to download a missing system file.

For now there is no listing option provided, but this will come later.  Closes #5742.

PaulWessel commented:
I added a listing modifier +n to list available collections, i.e. 

`gmt coast -E+n`

If you give a country or countries it will only list collections that contains that country:

`gmt coast -ENO+n`

should find Scandinavia, Nordic Countries, etc. This assume there are such collections of course

Esteban82 commented:
Great Paul

BTW, I try to use "" for the names 
The DCW file:
```
# Central America
tag: CAM "CentralAmerica"
BZ,CR,HN,GT,NI,PA,SV

gmt coast -R"CentralAmerica" -Gred -B -png mapa
coast [WARNING]: No country code matching CentralAmerica (skipped)

```

PaulWessel commented:
Try no quotes inside the file, and only use quotes if there are spaces on the command line.

Esteban82 commented:
> Try no quotes inside the file, and only use quotes if there are spaces on the command line.

Great. Now works.


For the record:

```
# Central America
tag: CAM Central America
BZ,CR,HN,GT,NI,PA,SV

gmt coast -R"Central America" -Gred -B -png mapa
```

Esteban82 commented:
So, this seems to work fine. 

Should I approve the PR? 
Where should I write the list for dcw-collections.txt?
What about other regions which do not include countries like seas?

Also, I was thinking that maybe we could add the full name for countries (e.g. `-RJapan`).

PaulWessel commented:
> Should I approve the PR? 

Not yet. We will wait until the dcw-collections.txt have been reviewed by the team.

> Where should I write the list for dcw-collections.txt?

That file should be added to the dcw-gmt repo where the country and state lists are.

> What about other regions which do not include countries like seas?

Those regions are more like named region with a corresponding -Rw/e/s/n string and does not have anything to do with DCW per se.  Not sure, but it would be a different file, such as _gmt-regions.txt_ with records like
```
# Named oceans
w/e/s/n SRG Sargasso Ocean
w/e/s/n NATL North Atlantic Ocean
etc.
```

But maybe we do add it to DCW so that **-E+n** can list them as well - they are just different in that it is just a shorthand for a w/e/s/n setting, but the padding via **+r**|**R** still applies to those regions.  So make a separate _dcw-regions.txt_ file for dcw-gmt repo instead.

> 
> Also, I was thinking that maybe we could add the full name for countries (e.g. `-RJapan`).

OK, those names are already in the DCW tables so this is a software update in GMT only,

anbj commented:
Just a thought: this feature is now hard linked to `DCW`, but it need not be. The feature of defining areas of interests is very handy, regardless of `DCW`. `DCW` makes it super convenient in that you can supply a bunch of country codes. It should also be possible to being able to define an `-R`

Like:

```
0/360/55/90 IBCAO International bathy.....
```

Edit: And I see Paul has commented on this in https://github.com/GenericMappingTools/gmt/pull/6064#issuecomment-981249640.

anbj commented:
Minor comment: Ability to separate blocks with an empty line - pure cosmetical/convenience. Especially nice if it's possible to define `-R` with no country codes (to be able to easilty read the file):

```
$ cat .gmt/geography/dcw/dcw-collections.txt 
tag: SCA Scandinavia
NO,SE,DK

tag: NOR Fullnametest
NO
```
```
$ gmt pscoast -E+n
pscoast [ERROR]: Collection file /home/anbj/.gmt/geography/dcw/dcw-collections.txt is not in correct format - exit
-R0/0/0/0
```

Alright - remote empty line and check again:

```
$ cat .gmt/geography/dcw/dcw-collections.txt 
tag: SCA Scandinavia
NO,SE,DK
tag: NOR Fullnametest
NO
```

The `tag` does not seem to be correctly read for the second block ("Fullnametest"). Both turn up as `SCA`:
```
$ gmt pscoast -E+n
Scandinavia	SCA	NO,SE,DK
Fullnametest	SCA	NO
```

anbj commented:
How is the user supposed to add entries? Edit `dcw-collections.txt` directly? How does `gmt` handle this? I guess the file is checked against a hash?

PaulWessel commented:
The dcw-collection is read only for users. They instead edit a dcw-groups.txt file in their ./gmt directory

anbj commented:
Ok, so the dcw-collection is made by the devs with a set of assumed relevant/interesting areas.

Esteban82 commented:
> Ok, so the dcw-collection is made by the devs with a set of assumed relevant/interesting areas.

Yes, that is my idea. The user then, can modified or add the records on that file.

About the list, I think we should follow some guidelines. I am thinking on:

1. Collections of countries that are "stable" on time. So they should be based on cultural, historical and geographic aspects and maybe not in political or economical (e.g. Eurozone) ones. The later could change with time and we should update the list. 
2. We should use international and well-known regions. Maybe use collections define by an organism like U.N.?
3. Exclude countries that are partially identified for a collection. For example, Mexico sometimes is defined as part of Central America and sometimes not. I think we should exclude because we always could do `-R"Central America",MX`


Some doubts: 
1. Should we limit the number of countries per collection? i.e. only define collections which has at least 4countries? The users can easily handle with 3 countries. Not sure if the limit should be 4. Maybe. 
2. We should limit the extend of collections?

PaulWessel commented:
Just so we are clear, there are for now two separate files: dcw-collections.txt comes with the DCW installation and is a list of things we developers wishes for all to have access to.  Then, users can extend that reach by adding their own dci-groups.txt file to ~/.gmt.  We can discuss if both files should be called dcw-collections.txt which may be more consistent.

There seems also to be a need for a list of named regions that are shorthands to specific -Rw/e/s/n settings. We can worry later if those should also go into the same file or a separate file.  I can code it either way.  But likewise, we should supply some basic well known cases in the DCW repo distribution and then users can create their own extensions to that.

Esteban82 commented:
Here there is the sub-regions use by the United Nations:

https://en.wikipedia.org/wiki/United_Nations_geoscheme.

PaulWessel commented:
> Here there is the sub-regions use by the United Nations:
> 
> https://en.wikipedia.org/wiki/United_Nations_geoscheme.

Yes, that list plus your 9-10 oceanic regions would be a good set of named regions for DCW to provide.  Then users can add odd regions are they see fit.


So, since we don't want the users to deal with too many files, perhaps both the current tag and the naemd region should be in a single file with mix and match.  Perhaps this is the format of _dcw-collections.txt_:

```
# Central America
tag: CAM Central America
list: BZ,CR,HN,GT,NI,PA,SV
# The Sargasso sea
tag: SRG Sargasso Sea
region: -88/-66/12/35
...
```

That means we have ~30 named entries supplied via DCW and the E+n will list them.  I can make the code changes later today, including skipping blank lines as per @anbj.

PaulWessel commented:
Update: Here are some changes:

1. The user file and the DCW system file now have the same name dcw-collections.txt
2. The file now supports both the collection lists and name region wesn as suggsted above
3. I fixed the bug that report the same code for all collections
4. I added some documentation to the coast man page
5. I tested this with a dummy north sea region and combinations like -RNO,NSEA worked fine.
6. The name regions do not correspond to any data so you cannot use -RNorthSea to extract data.  This may lead to a crash - have not checked yet.

Esteban82 commented:
> Yes, that list plus your 9-10 oceanic regions would be a good set of named regions for DCW to provide. Then users can add odd regions are they see fit.

I see that the UN region are follow the below code where each region is defined by 3 numbers.
 https://en.wikipedia.org/wiki/UN_M49

So we get this for example:
```
# Central America
tag: 029 Central America
list: BZ,CR,HN,GT,NI,PA,SV

```
Also, there are some major sub-regions. For example sub-saharan africa which includes four sub-regions (western africa, central africa, southern africa, eastern africa).  Another major sub-region is Latin American and the Caribbean. I am not including them,

PaulWessel commented:
Do we want to use UN029 instead of 029? I worry that giving an integer will cause trouble with the -R parsing.  Currently it checks for letters.

Esteban82 commented:
Yes, I will add it. I get an error for -R (but not for -E).

anbj commented:
1. What about adding a header when listing groups/regions? Something like:
```
$ gmt pscoast -E+n
Description/Name              Tag          Country code(s)/region                 <---------
FrancoItalic Union	FRIT	FR,IT
Sargasso Sea	SARG	70W/40W/20N/35N
```

2. Formatting output from `gmt pscoast -E+n` should ideally line up nicely, but this may be difficult due to variable length of the different fields (especially the "Description/Name"). E.g. a tag including two country codes vs. a tag including 10 country codes. Of these, the "Description/Name" column may potentially be the messiest - output that as the 3. column?

3. It seems that country codes in general are case sensitive (at least the first letter -  see below). Should they be? If it required, it should be noted in the man-pages. Case is also important for this new feature:

If I create a tag with lowercase:

```
# Stay away from those eels!
tag: sarg Sargasso Sea
region:  70W/40W/20N/35N
```

```
$ gmt pscoast -Esarg
pscoast [ERROR]: Could not parse sarg into Geographical, Cartesian, or Temporal coordinates!
pscoast [ERROR]: Option -R parsing failure. Correct syntax:
[...]
```

Change the tag's first letter to uppercase (but don't change tag in `dcw-collections.txt`). Note different warning compared to the example above:

```
$ gmt pscoast -ESarg
pscoast [WARNING]: No country code matching Sarg (skipped)
pscoast [WARNING]: Column selected for latitude-formatting has values that exceed +/- 90; set to NaN
pscoast [WARNING]: Column selected for latitude-formatting has values that exceed +/- 90; set to NaN
-R0/0/NaN/NaN
```

Change tag (in `dcw-collections.txt`) as well:

```
# Stay away from those eels!
tag: Sarg Sargasso Sea
region:  70W/40W/20N/35N
```

And it works fine:

```
 gmt pscoast -ESarg
-R-70/-40/20/35
```

So seems that it is important that the first letter is uppercase. If you change one of the other letters to uppercase you get trouble:

```
# Stay away from those eels!
tag: sarG Sargasso Sea
region:  70W/40W/20N/35N
```

```
$ gmt pscoast -EsarG
pscoast [ERROR]: Could not parse sarG into Geographical, Cartesian, or Temporal coordinates!
pscoast [ERROR]: Option -R parsing failure. Correct syntax:
[...]
```

Starting to nitpick now.

4. Use with `-R` seems to work:

```
tag: NOS Norwegian Sea
region: -25/20/60/75
```

Both

```
$ gmt pscoast -R"Norwegian Sea" -M -W |head
> Shore Bin # 37, Level 1
> Shore Bin # 37, Level 1
> Shore Bin # 37, Level 1
> Shore Bin # 37, Level 1
19.0000762951	74.3678950179
19.2561226825	74.4333562219
[...]
```

and 

```
gmt pscoast -R"Norwegian Sea" -JM5c -W -pdf test
```

runs fine (but same issue with letter case as descibed above).

PaulWessel commented:
Thanks, after my next zoom I will

1. Add a header and restructure how the output looks like.
2. Document that the tags must be upper case and that the full names must start with a Capital letter.  I can add checks for that during the parsing of the files.

I may need to revert back to two separate file names for the DCW system and the user file.  This has to do with our functions that searches the various paths and in my tests I found the same file twice.  Separate file names will fix this.

PaulWessel commented:
Some changes:

1. The user-controlled file is now called ~/.gmt/dcw.conf
2. I have added documentation to coast.rst regarding the format and specifications of that file
3. I have added more checks during file parsing to catch any errors in the data.

joa-quim commented:
> Document that the tags must be upper case and that the full names must start with a Capital letter. I can add checks for that during the parsing of the files.

Why not accept case insensitive tags? So much simpler for the users.

PaulWessel commented:
> > Document that the tags must be upper case and that the full names must start with a Capital letter. I can add checks for that during the parsing of the files.
> 
> Why not accept case insensitive tags? So much simpler for the users.

Maybe, but harder for the developer who now have to check other things.

PaulWessel commented:
OK, bending to the will of others: While the ISO 2-char codes remain upper case, the custom collections and named region can be specified in lower case, but not in the input dcw.conf file.  I.e., that file must have things like SCA Scandinavia but the user can use -Rsca or -Rscandinavia or -RScAnDiNaViA as they wish.

Esteban82 commented:
For the oceans, I suggest to use this info which includes a shp where "the limits of seas were merged with the shoreline of the GSHHS database (Wessel and Smith 1996), and names are those of the IHO 1953 nomenclature." 

There area around 100 seas and oceans. If approve, we just have to extract the region for each one. 


> Conditions of use
> 
> This map is freely available for non-commercial use.
> It is registered with the French Agence pour la Protection des Programmes (APP) and has the following Inter Deposit Digital Number (IDDN): IDDN.FR.001.500006.000.R.C.2011.000.40000.

https://www6.rennes.inrae.fr/ese_eng/HRMLOS

anbj commented:
Great Paul!

> OK, bending to the will of others: While the ISO 2-char codes remain upper case, the custom collections and named region can be specified in lower case, but not in the input dcw.conf file. I.e., that file must have things like SCA Scandinavia but the user can use -Rsca or -Rscandinavia or -RScAnDiNaViA as they wish.

Just out of curiosity; why is case important in the `~/.gmt/dcw.conf`? Lower or upper case og mix; why does it matter? Ref. joa-quim's remark.

PaulWessel commented:
I guess I just don't like telling people to define abbreviations and then let those be lower case.  I guess I instantly would make those upper case in any custom group I make myself.  But, imposing this restriction on others is not really important to me so I will remove the checks on this.

PaulWessel commented:
> For the oceans, I suggest to use this info which includes a shp where "the limits of seas were merged with the shoreline of the GSHHS database (Wessel and Smith 1996), and names are those of the IHO 1953 nomenclature."
> 
> There area around 100 seas and oceans. If approve, we just have to extract the region for each one.

You mean to find the w/e/s/n that encloses each of these polygons, right?

PaulWessel commented:
Adding @meghanrjones to this PR so she is aware.  These features are dormant if DCW 2.0.1 is used but users with master (once this branch is merged) can take advantage of making their own collections and named regions, and once DCW 2.0.2 is released it will come with the regions/collections discussed above that we provide.

Esteban82 commented:
> You mean to find the w/e/s/n that encloses each of these polygons, right?

Yes, exactly. must use gmt convert to extract each polygon in a separate file and then gmt info to get the region, right?

PaulWessel commented:
If the shape file has all the polygon in one place , yes that would be the safest way.  Thanks!
I will be working in GMT to relax the ISO code stuff so that one could also give -RFrance instead of just -RFR, since we are allowing full names for other things.

Esteban82 commented:
> I will be working in GMT to relax the ISO code stuff so that one could also give -RFrance instead of just -RFR, since we are allowing full names for other things.

Great. It would be available also for continents? i.e. to do `-RAfrica`?

PaulWessel commented:
Yes, I can do continents too.

Esteban82 commented:
> gmt convert to extract each polygon in a separate file a

Paul, I have some difficulties with this step. I do `gmt convert `and I get a file with all the data but without the name of each sea. Do you know how can I get each segment and keep its name?

PaulWessel commented:
Yes, I see.  After conversion to OGR_GMT there are 105 records with `@D` statements that has the names.  However, there are lots of other segments for holes so there are 180115 segments in total.   `gmt info -As -Ie HRmLOS_1.1.gmt` gives the -R string for all of those segments but you just want 105 of those since the rest are holes. But gripping on `@P` gives 111 polygons so some seas are split into two pieces...
I don't think GMT can do what we want here.  Perhaps better to have a loop that extracts one multipolygon at the time from the shape file and then examine just one at the time.  Then it is easier to grep out the name and fine the region?

PaulWessel commented:
BTW, now the -R and -E options can handle continent names and country names and state names:

```
gmt coast -RUS -JM15c -ETexas+gred -ECalifornia+gblue -W -B -pdf t
gmt pscoast -ENorway,France
gmt pscoast -EUtah
```

joa-quim commented:
Men of no faith and bash lovers

```
julia> HRML = gmtread("C:\\v\\HRMLOS_shapefile\\HRmLOS_1.1.shp");

julia> HRML[1].attrib
Dict{String, String} with 5 entries:
  "IHO1953_Nb" => "ND"
  "Sph_Area"   => "20248553745600.00000000000"
  "IHO53_Name" => "Southern Ocean"
  "Parts"      => "1.00000000000"
  "Sph_Len"    => "62709773.23809999973"

# Have to use a try catch here because apparently in the shape some polygons lack attributes
julia> for k = 1:length(HRML) try println(HRML[k].attrib["IHO53_Name"]); catch end end

Southern Ocean
Southern Ocean
Bass Strait
Bass Strait
Rio de La Plata
Rio de La Plata
Tasman Sea
Tasman Sea
Great Australian Bight
Great Australian Bight
Savu Sea
Savu Sea
Timor Sea
Timor Sea
Mozambique Channel
Mozambique Channel
Bali Sea
Bali Sea
Coral Sea
Coral Sea
Flores Sea
Flores Sea
Solomon Sea
Solomon Sea
Arafura Sea
Arafura Sea
Gulf of Boni
Gulf of Boni
Ceram Sea
Ceram Sea
Bismarck Sea
Bismarck Sea
Banda Sea
Banda Sea
Java Sea
Java Sea
Gulf of Tomini
Gulf of Tomini
Singapore Strait
Singapore Strait
Makassar Strait
Makassar Strait
ND
ND
South Pacific Ocean
South Pacific Ocean
Halmahera Sea
Halmahera Sea
South Atlantic Ocean
South Atlantic Ocean
Molukka Sea
Molukka Sea
Celebes Sea
Celebes Sea
Malacca Strait
Malacca Strait
ND
ND
Indian Ocean
Indian Ocean
Sulu Sea
Sulu Sea
Laccadive Sea
Laccadive Sea
Andaman Sea
Andaman Sea
Caribbean Sea
Caribbean Sea
Gulf of Guinea
Gulf of Guinea
Gulf of Aden
Gulf of Aden
Gulf of Thailand
Gulf of Thailand
Gulf of Aqaba
Gulf of Aqaba
Gulf of Suez
Gulf of Suez
Gulf of Oman
Gulf of Oman
ND
ND
Eastern China Sea
Eastern China Sea
Seto naikai
Seto naikai
Philippine Sea
Philippine Sea
Strait of Gibraltar
Strait of Gibraltar
Alboran Sea
Alboran Sea
Red Sea
Red Sea
South China Sea
South China Sea
Mediterranean Sea - Eastern Basin
Mediterranean Sea - Eastern Basin
Ionian Sea
Ionian Sea
Sea of Marmara
Sea of Marmara
Gulf of Mexico
Gulf of Mexico
Balearic
Balearic
Tyrrhenian Sea
Tyrrhenian Sea
Bay of Fundy
Bay of Fundy
Mediterranean Sea - Western Basin
Mediterranean Sea - Western Basin
Aegean Sea
Aegean Sea
Ligurian Sea
Ligurian Sea
Bay of Biscay
Bay of Biscay
Arabian Sea
Arabian Sea
Adriatic Sea
Adriatic Sea
English Channel
English Channel
Gulf of Iran
Gulf of Iran
Gulf of St. Lawrence
Gulf of St. Lawrence
Gulf of California
Gulf of California
Celtic Sea
Celtic Sea
Bristol Channel
Bristol Channel
Black Sea
Black Sea
Sea of Azov
Sea of Azov
Irish Sea and St. George's Channel
Irish Sea and St. George's Channel
Bay of Bengal
Bay of Bengal
ND
ND
Yellow Sea
Yellow Sea
North Pacific Ocean
North Pacific Ocean
Inner Seas of the West Coast of Scotland
Inner Seas of the West Coast of Scotland
Kattegat, Sund and Belts
Kattegat, Sund and Belts
Gulf of Riga
Gulf of Riga
Japan Sea
Japan Sea
Labrador Sea
Labrador Sea
Baltic Sea
Baltic Sea
North Sea
North Sea
Skaggerak
Skaggerak
Coastal Waters of Southeast Alaska and British Columbia
Coastal Waters of Southeast Alaska and British Columbia
Gulf of Finland
Gulf of Finland
Hudson Strait
Hudson Strait
Gulf of Alaska
Gulf of Alaska
Sea of Okhotsk
Sea of Okhotsk
Hudson Bay
Hudson Bay
Gulf of Bothnia
Gulf of Bothnia
Bering Sea
Bering Sea
White Sea
White Sea
Davis Strait
Davis Strait
Chukchi Sea
Chukchi Sea
North Atlantic Ocean
North Atlantic Ocean
Beaufort Sea
Beaufort Sea
Norwegian Sea
Norwegian Sea
East Siberian Sea
East Siberian Sea
Laptev Sea
Kara Sea
Barentsz Sea
The Northwestern Passages
Baffin Bay
Greenland Sea
Lincoln Sea
Arctic Ocean
```

PaulWessel commented:
Making that list is easy, even for bash-lovers.  But isolating each segment which may be two segments and get their w/e/s/n is not directly possible.  For instance, the number of polygons exceed the list of names so some names have two polygons, i.e., disjointed multi polygon.  GMT has no way of representing that as one segment.  Not sure if your Julia implementation does?

joa-quim commented:
Apparently the second polygon is the one that matter for the zone definition and the one that has a BoundingBox. Unclear right now why the first one (with coast lines) lacks it.
```
julia> HRML[1].bbox
Float64[]

julia> HRML[2].bbox
4-element Vector{Float64}:
 -46.18872199999993
 -46.14372199999991
 -60.51722199999995
 -60.49249999999995

julia> size(HRML[1])
(17651, 2)

julia> size(HRML[2])
(9, 2)
```

PaulWessel commented:
It would be great if you could help @Esteban82 and provide a list of names and corresponding w/e/sn so we can wrap up DCW 2.0.2.

PaulWessel commented:
@meghanrjones, I think this PR is complete as far as the GMT coding and documentation updates go. For testing one can just stick a few lines of collections into a _dcw-collections.txt_ file and place it in the DCW directory, then add a _dcw.conf_ file in ~/.gmt for testing.

joa-quim commented:
Ghrr, more debugging for me. ``gmtread`` is much faster but have some shits. When reading with GDAL I get
```
julia> HRML = gdalread("C:\\v\\HRMLOS_shapefile\\HRmLOS_1.1.shp");          # Damn slow .... but
```
there are a couple of ``NS``'s
```
julia> for k = 1:length(HRML) println(HRML[k].attrib["IHO53_Name"])  end
Southern Ocean
Bass Strait
Rio de La Plata
Tasman Sea
Great Australian Bight
Savu Sea
Timor Sea
Mozambique Channel
Bali Sea
Coral Sea
Flores Sea
Solomon Sea
Arafura Sea
Gulf of Boni
Ceram Sea
Bismarck Sea
Banda Sea
Java Sea
Gulf of Tomini
Singapore Strait
Makassar Strait
ND
South Pacific Ocean
South Pacific Ocean
Halmahera Sea
South Atlantic Ocean
Molukka Sea
Celebes Sea
Malacca Strait
ND
Indian Ocean
Sulu Sea
Laccadive Sea
Andaman Sea
Caribbean Sea
Gulf of Guinea
Gulf of Aden
Gulf of Thailand
Gulf of Aqaba
Gulf of Suez
Gulf of Oman
ND
Eastern China Sea
Seto naikai
Philippine Sea
Strait of Gibraltar
Alboran Sea
Red Sea
South China Sea
Mediterranean Sea - Eastern Basin
Ionian Sea
Sea of Marmara
Gulf of Mexico
Balearic
Tyrrhenian Sea
Bay of Fundy
Mediterranean Sea - Western Basin
Aegean Sea
Ligurian Sea
Bay of Biscay
Arabian Sea
Adriatic Sea
English Channel
Gulf of Iran
Gulf of St. Lawrence
Gulf of California
Celtic Sea
Bristol Channel
Black Sea
Sea of Azov
Irish Sea and St. George's Channel
Bay of Bengal
ND
Yellow Sea
North Pacific Ocean
North Pacific Ocean
Inner Seas of the West Coast of Scotland
Kattegat, Sund and Belts
Kattegat, Sund and Belts
Gulf of Riga
Japan Sea
Labrador Sea
Baltic Sea
North Sea
Skaggerak
Coastal Waters of Southeast Alaska and British Columbia
Gulf of Finland
Hudson Strait
Gulf of Alaska
Sea of Okhotsk
Hudson Bay
Gulf of Bothnia
Bering Sea
Bering Sea
White Sea
Davis Strait
Chukchi Sea
Chukchi Sea
North Atlantic Ocean
Beaufort Sea
Norwegian Sea
East Siberian Sea
East Siberian Sea
Laptev Sea
Kara Sea
Barentsz Sea
The Northwestern Passages
Baffin Bay
Greenland Sea
Lincoln Sea
Arctic Ocean
```
and the BoundingBoxes

```
julia> for k = 1:length(HRML) println("-R",HRML[k].bbox[1],"/",HRML[k].bbox[2],"/",HRML[k].bbox[3],"/",HRML[k].bbox[4])  end
-R-179.99999999999991/180.0/-78.57036099999993/-59.99999999999994
-R143.5167371450001/149.91503765900006/-41.260416999999904/-37.48877799999991
-R-58.63544399999995/-54.942110999999954/-36.37047199999995/-33.06713899999994
-R146.82888235200005/175.2717239760001/-50.880443999999954/-29.999999998999954
-R117.6312220000001/146.23044400000003/-43.55777439499991/-31.464582999999948
-R118.45945781500006/125.0/-10.93735909599991/-8.261249999999961
-R122.8456713380001/132.76294400000006/-15.517110999999943/-8.173545154999942
-R32.49866700000007/49.24122200000005/-26.845900539999946/-10.504848318999905
-R114.33788900000002/117.18442326400009/-9.00781528799996/-7.008998508999923
-R141.04203900100003/170.18639764800002/-29.999999998999954/-7.387943999999948
-R117.16646259800007/122.94294400000001/-8.745443999999907/-5.511249999999961
-R146.94544400000007/162.355230161/-11.836072324999918/-4.834415971999931
-R130.76628020700002/142.16375000000005/-17.72713899999991/-2.910443999999927
-R120.18544400000007/121.90482906900002/-5.619762350999906/-2.621249999999918
-R125.91830964100006/133.99794400000007/-5.278520227999934/-1.177449835999937
-R141.99990141/152.8762220000001/-5.965443999999934/-1.0702496209999595
-R120.46832754700006/133.14520997600005/-8.905388999999957/-0.8362779999999361
-R104.51627800000006/119.47849503900011/-7.811737184999913/-2.576086559999908
-R120.00208300000008/123.41097878200003/-1.4329999999999359/0.5512220000000525
-R103.34204271300007/104.57041700000008/1.0736670000000572/1.4295280000000616
-R115.96463900000003/120.83536100000003/-5.599061945999949/1.3185601960000781
-R103.62621804600008/104.13875000000007/1.3219834730000457/1.6628890000000638
-R-179.99999999999991/-67.26666667599994/-59.99999999999994/1.9000597870000888
-R130.1054170000001/180.0/-59.99999999999994/3.6997398850000422
-R127.54280041600009/130.93211100000008/-1.6429595219999555/2.4925379830000907
-R-69.58586099999991/20.000053691000062/-59.99999999999994/5.684341886080802e-14
-R123.31287738700007/128.562664874/-1.8449489289999406/4.480397046000064
-R117.01372200000003/125.60354206/0.8079170000000886/7.8528890000000615
-R95.45445983800005/103.51107589900005/0.8081660060000786/8.408722000000068
-R-80.09149899999994/-79.55731307499991/8.926062379000086/9.332507000000078
-R20.000000000000057/146.836278/-59.99999999999994/10.437097927000082
-R116.73625000000004/123.42875000000004/5.018657663000056/13.44166203800006
-R71.67764964900005/80.58788900000008/-0.6973484469999107/14.859972000000084
-R92.40124484200004/98.88541700000007/5.5393201610000915/17.509472000000073
-R-88.93694399999993/-59.38176120799994/7.904917000000069/22.705444000000057
-R-7.593319901999905/9.987083000000041/-0.8521109999999226/6.621250000000089
-R42.51625000000007/51.27661016500008/10.375389000000041/15.213720196000054
-R99.14041700000007/105.11501624700003/6.167833000000087/13.608750000000043
-R34.365215127000056/35.00275000000005/27.91479122000004/29.54944400000005
-R32.34205600000007/34.24536100000006/27.45496976700008/29.959583000000066
-R56.26627800000006/61.78950000000003/22.514556000000084/25.940306000000078
-R32.25976500000007/32.588550000000055/29.932713946000035/31.259760000000085
-R119.4437220000001/131.12215282900002/24.05392217900004/33.29955600000005
-R130.74461100000008/135.46622200000002/33.23291700000004/34.80622200000005
-R120.09751845700009/147.79204595400006/2.4987500000000864/35.14638722300009
-R-6.031920857999921/-5.293426278999959/35.77869400000009/36.18708300000009
-R-5.357999999999947/-1.1929439999999545/35.072861000000046/36.84038900000007
-R33.637057248000076/43.49783300000007/12.457028000000037/28.138750000000073
-R102.24491755000008/122.16451858200003/-3.2323274109999147/25.619556000000046
-R10.015361000000041/36.209250000000054/30.266222000000084/37.798722000000055
-R15.082944000000055/23.225389000000064/36.402493080000056/40.520417000000066
-R26.176770244000068/29.939583000000084/40.00069400000007/41.22886768700005
-R-97.89111099999991/-80.37205599999993/18.06080600000007/30.775056000000063
-R-0.34130599999991773/4.289625557000079/38.649050903000045/41.89205600000008
-R8.648750000000064/16.221250000000055/37.80400899300008/44.10958300000004
-R-67.2162219999999/-63.37724999999995/44.09651570400007/45.975944000000084
-R-2.1827932689999443/12.423311373000047/35.57288900000009/43.784583000000055
-R22.527028000000087/28.335417000000064/35.11036100000007/41.012056000000086
-R7.5326687730000685/9.848041078000051/43.00955600000009/44.42711100000008
-R-7.877943999999957/-0.5071389999999383/43.275389000000075/47.91294400000004
-R51.02205600000008/74.10999663200005/-0.6912779999999543/25.625417000000084
-R12.141194000000041/20.017833000000053/39.743266507000044/45.79455600000006
-R-5.701737266999942/1.7171444490000454/48.49619400000006/51.13877590000004
-R47.702889000000084/57.34119400000009/23.90655600000008/30.514556000000084
-R-66.79549999999995/-54.7037499999999/45.32293281400007/52.22122200000007
-R-114.9551669999999/-106.79586099999995/22.880667000000074/31.884306000000038
-R-11.499999998999954/-4.157499999999914/46.499938960000065/52.26372200000009
-R-4.9268022099999484/-2.3475279999999543/50.987944000000084/51.81708300000008
-R27.306167000000073/41.779528000000084/40.908750000000055/46.889556000000084
-R34.46544400000005/39.305611000000056/45.10262032800006/47.28955600000006
-R-6.5075279999999225/-2.680860999999936/51.901984103000075/55.007889000000034
-R78.90038900000008/95.02801177900005/5.758601628000065/22.77033300000005
-R-71.21002799999991/-64.49999999999994/46.812722000000065/50.302917000000036
-R117.545389/126.99372200000005/33.29050940700006/41.048722000000055
-R128.68368435800005/180.0/5.684341886080802e-14/56.25197200000008
-R-179.99999999999991/-77.02263899999991/5.684341886080802e-14/58.19795555600007
-R-8.282704592999949/-4.427527999999938/54.345781196000075/58.62272739000008
-R11.870833000000061/12.19566625300007/54.843585956000084/55.00458300000008
-R9.420778000000041/13.069139000000064/54.63392522700008/58.349556000000064
-R21.768573597000056/24.589083000000073/56.96111100000007/59.22796002200005
-R125.80325137300008/142.25583300000005/32.57838602500004/51.73925752800005
-R-64.3026109999999/-43.69162243899996/47.38644400000004/60.35288900000006
-R9.840806000000043/23.50598146100009/53.601639000000034/59.93622200000004
-R-4.444193999999925/9.835806000000048/50.93373021900004/61.02791700000006
-R7.06744400000008/11.705694000000051/57.10288900000006/59.91955600000006
-R-137.1464999999999/-122.17802799999993/47.036366000000044/59.481750000000034
-R22.897084229000086/30.347083000000055/59.21536100000009/60.742833000000076
-R-80.9547645959999/-64.43274627699992/58.00194400000004/64.90275000000008
-R-163.37588899999992/-136.61752799999994/54.29193171600008/61.52219400000007
-R135.17080600000008/165.308583/43.21455600000007/62.69458300000008
-R-94.9725279999999/-76.51936099999995/51.132472000000064/66.05580600000008
-R17.039083000000062/25.509167000000048/59.620299156000044/65.90708300000006
-R161.8907220000001/180.0/51.10000000000008/65.03377800000004
-R-179.99999999999991/-156.84166699999992/51.12350092000008/66.54202800000007
-R31.869278000000065/44.58675000000005/63.78144400000008/68.65204469200006
-R-70.0622219999999/-44.46591699999993/59.89288900000008/70.07291700000007
-R177.490423847/180.0/68.99377800000008/70.97294400000004
-R-179.99999999999991/-156.4813609999999/65.95822200000003/71.80722944700005
-R-83.01544399999995/6.667167084000084/5.684341886080802e-14/68.64455600000008
-R-156.6413609999999/-122.60438899999991/68.63991700000008/76.35110697900006
-R-13.497455457999934/25.76914606200006/60.86205600000005/76.56275526200005
-R-179.99999999999991/-178.5952742179999/71.52708300000006/72.10442831000006
-R138.92590996600006/180.0/68.76369400000004/77.12507914800005
-R95.70283300000006/141.30877800000007/70.68036100000006/81.23303782200003
-R55.08111100000008/104.69841214700006/66.19500000000005/81.45670855600008
-R16.581500000000062/68.55576621200004/66.81647200000003/81.82700060000008
-R-128.03154733899993/-72.04719399999993/63.78413900000004/81.70261100000005
-R-82.21997199999993/-50.52124999999995/69.61669400000005/82.45779076800005
-R-36.88074999999992/17.398389000000066/65.08625000000006/83.58119821900004
-R-69.82717446299995/-34.467371457999946/81.60350000000005/83.73493125200008
-R-179.99999999999991/180.0/71.38612125700007/90.00000000000006
```

PaulWessel commented:
Thanks @joa-quim !  OK, @Esteban82 I think you can work with this instead, yes?

PaulWessel commented:
One stupid thing is this:

```
Mediterranean Sea - Eastern Basin
Mediterranean Sea - Western Basin

```
So we will have to manually add a separate Mediterranean Sea for the whole enchilada.

Esteban82 commented:
> Thanks @joa-quim !  OK, @Esteban82 I think you can work with this instead, yes?

Yes. That looks very helpful. Just one doubt. How should I report the region when the sea goes through -180 longitude? Should I use longitudes from 0 to 360?

PaulWessel commented:
Does not matter.  Since you type strings you can do 140E/170W or 140/190 I think.

Esteban82 commented:
> -R-179.99999999999991/180.0/-78.57036099999993/-59.99999999999994
> -R143.5167371450001/149.91503765900006/-41.260416999999904/-37.48877799999991

@joa-quim it would be possible to have the region list with few decimal numbers (5 or 6)?  I try those but I got this error which dissapear when I reduce the decimal numbers:
```

coast [ERROR]: Could not decode -58.635443999999-54.942110999999-36.370471999999-33.06713899999994, return NaN.
ERROR: Caught signal number 11 (Segmentation fault) at
/lib/x86_64-linux-gnu/libc.so.6(+0xb4dc4)[0x7f29c9165dc4]
[0xffffffc0]
```

PaulWessel commented:
That should not matter.  What command do you run when this breaks?

Esteban82 commented:
`gmt coast -Bagf -Ggray -png map -N1 -RRDP -JM15c`

And the files for dcw-collections was
It works:
```
tag: RDP Rio de La Plata
region: -58.63544/-54.94211/-36.370472/-33.06714
```


It brakes (original):

```
tag: RDP Rio de La Plata
region: -58.63544399999995/-54.942110999999954/-36.37047199999995/-33.06713899999994
```

With this also works (I just deleted the last 3 numbers):

`region: -58.63544399999995/-54.942110999999954/-36.37047199999995/-33.06713899999`

PaulWessel commented:
Just pull and rebuild and it should work.  I only used 16 char per coordinate and this exceeded it.

Esteban82 commented:
Now it is working. I think tomorrow I will have the final regions

Any suggestions on how to define the codes for the seas? I though in 2 options. 

1. We create them. Maybe we could use 3 letters for the name and 1 for the type of water body (sea, ocean, bay, strait, etc).
2.  In the file the seas have also a number (IHO1953_Nb). Maybe I could use them although there are some seas that do not have (maybe it is just an error of the shp file. I can check on internet). This code is two number and sometimes a letter. https://es.wikipedia.org/wiki/Anexo:Oc%C3%A9anos_y_mares_del_mundo

PaulWessel commented:
Would be best if we did not have to create anything but use official things, if possible.

Esteban82 commented:
> Would be best if we did not have to create anything but use official things, if possible.

Ok. I will use IHO them. 

If we stick to the official IHO1953 them I should remove the Southern Ocean (which any user can easily do). 

For the record I will append IHO to the tags. For example:

```
tag: IHO33 Rio de La Plata
region: -58.63544/-54.94211/-36.370472/-33.06714

```

PaulWessel commented:
Why skip southern ocean? No IHO code for it?

Esteban82 commented:
Exactly. No code. Although it is the ocean south of -60. - > So any user can use -R-180/180/-80/-60.

Esteban82 commented:
I have to doubts/Issue with the list of oceans:

1. How should I name the codes with one number? IHO01 or IHO1?

2. In the original document (IHO1953) there are seas and, for in some case, there are internal sub-divisions. For example the Baltic sea (sea number 1) includes the Gulf of Bothnia (1a), Finland (1b) and Riga (1c). However, in the shp file those polygons are exclude from the main sea. To be clear, this is map of the Baltic Sea. It doesn't include the gulfs.

![map](https://user-images.githubusercontent.com/17908951/144536548-8bbcc761-eae0-418a-97d2-2160e6dd1951.png)

What should I do? Should I include the inner divisions, right? Is there any trick to get the whole region from the data that I have? 


```
tag: IHO1 Baltic Sea
region: 9.840806000000043/23.50598146100009/53.601639000000034/59.93622200000004
tag: IHO1a Gulf of Bothnia
region: 17.039083000000062/25.509167000000048/59.620299156000044/65.90708300000006
tag: IHO1b Gulf of Finland
region: 22.897084229000086/30.347083000000055/59.21536100000009/60.742833000000076
tag: IHO1c Gulf of Riga
region: 21.768573597000056/24.589083000000073/56.96111100000007/59.22796002200005
```

PaulWessel commented:
I think if you have 100+ IHO labels then use three digits with leading zeros, but if < 100 then one leading zero, so IHO01 I would suggest.

If I understand your other questions correctly you want maybe do something like this.  If the Baltic Sea really includes those gulfs, then I would do list this one after the others:

```
tag: HO1 Baltic Sea
list: IHO1a, IHO1b, IHO1c
```

Esteban82 commented:
> If I understand your other questions correctly you want maybe do something like this. If the Baltic Sea really includes those gulfs, then I would do list this one after the others:
> 
> ```
> tag: HO1 Baltic Sea
> list: IHO1a, IHO1b, IHO1c
> ```

Yes and no. For the Baltic Sea it doesn't work because there sum of those gulfs don't covered entirely Baltic Sea. There are areas that are outside those gulfs and do not have a particular name.
![imagen](https://user-images.githubusercontent.com/17908951/144596161-21e33754-d5f2-4b66-a0ea-abff3fe535bb.png)

However I thought in using that trick with other UN regions (with UN codes) like Latin American and the Caribbean and Sub-Saharan Africa.

Esteban82 commented:
I am getting this error when I add an extra record in the dcw-collections.txt. 

```
tag: IHO52 Japan Sea
region: 125.80325137300008/142.25583300000005/32.57838602500004/51.73925752800005

gmt coast -Bagf -Ggray -png mapa -N1 -JW15c -RIHO55
double free or corruption (out)
Abortado (`core' generado)

```
I suspect that the is somehow related to the number of records because the records that I add it is the number 101.

Esteban82 commented:
> One stupid thing is this:
> 
> ```
> Mediterranean Sea - Eastern Basin
> Mediterranean Sea - Western Basin
> ```
> 
> So we will have to manually add a separate Mediterranean Sea for the whole enchilada.

There is an issue here. In official documentation the western and eastern seas are defined.  So we should include them. 
https://epic.awi.de/id/eprint/29772/1/IHO1953a.pdf

So the dcw-collections.txt would be:

```
tag: IHO28 Mediterranean Sea
tag: IHO28A Mediterranean Sea Western Basin
tag: IHO28B Mediterranean Sea Eastern Basin
tag: IHO28a Strait of Gibraltar
tag: IHO28b Alboran Sea

```
When I used `-RIHO28A` I get the region of `-RIHO28a`. In the shp file the authors use 28-1 (instead of 28A) but I think we should stick to the original

PaulWessel commented:
OK, I think you should define the collections (Med W + Med E) using list: IHO28A, IHO28B.

Agree on sticking to the original and A or a.  I made the checks case insensitive but the _dcw-collections.txt_ file should stick to one syntax (I think I prefer the upper case for tags).

Esteban82 commented:
I got a segmentation fault error when I add a record like the following. It works if a remove the E and W in longitude values.


```
tag: IHO11 East Siberian Sea
region: 138.92590996600006E/178.5952742179999W/68.76369400000004/77.125079148

```

Full error message:
```
gmt coast -Bagf -Ggray -png Sea1 -N1 -JW15c -RIHO11
ERROR: Caught signal number 11 (Segmentation fault) at
/lib/x86_64-linux-gnu/libc.so.6(+0xb4dc4)[0x7f0f9ead8dc4]
[0xffffffc0]
Stack backtrace:
/usr/local/lib/libgmt.so.6(sig_handler_unix+0xf3)[0x7f0f9ed94bd3]
/lib/x86_64-linux-gnu/libc.so.6(+0x46210)[0x7f0f9ea6a210]
/lib/x86_64-linux-gnu/libc.so.6(+0xb4dc4)[0x7f0f9ead8dc4]
/lib/x86_64-linux-gnu/libc.so.6(+0x78eac)[0x7f0f9ea9ceac]
/lib/x86_64-linux-gnu/libc.so.6(+0x79f4d)[0x7f0f9ea9df4d]
/lib/x86_64-linux-gnu/libc.so.6(+0x8f11a)[0x7f0f9eab311a]
/usr/local/lib/libgmt.so.6(GMT_Report+0x1ae)[0x7f0f9ec8e55e]
/usr/local/lib/libgmt.so.6(+0x9ad19)[0x7f0f9ecd6d19]
/usr/local/lib/libgmt.so.6(gmt_DCW_operation+0x430)[0x7f0f9ecd7310]
/usr/local/lib/libgmt.so.6(gmt_parse_R_option+0x42e)[0x7f0f9edc432e]
/usr/local/lib/libgmt.so.6(gmt_parse_common_options+0x197b)[0x7f0f9edd61eb]
/usr/local/lib/libgmt.so.6(GMT_Parse_Common+0xbdb)[0x7f0f9ed9165b]
/usr/local/lib/libgmt.so.6(GMT_pscoast+0x1f9)[0x7f0f9efa9ec9]
/usr/local/lib/libgmt.so.6(GMT_coast+0x57)[0x7f0f9efb35b7]
/usr/local/lib/libgmt.so.6(GMT_Call_Module+0x3bf)[0x7f0f9ec98faf]
gmt(main+0x27a)[0x55855768b63a]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7f0f9ea4b0b3]
gmt(_start+0x2e)[0x55855768c80e]

```

PaulWessel commented:
While it is possible to use E and W suffices, you must make sure that west < east.  Here

`138.92590996600006E/178.5952742179999W`

is the same as

`138.92590996600006/-178.5952742179999`

We decided a long time ago that we should not "fix" this by adding 360 to east since then we could never catch any user errors.  So here you just have to use

`138.92590996600006E/181.404725782E`

There may be a few others but usually not many.

Esteban82 commented:
> So here you just have to use
> 
> 138.92590996600006E/181.404725782E

Ok. It is much easier to just add W.

PaulWessel commented:
Are there many of these? Since these are parsed in the dcw reader I could make an exception and add 360 when this happens.

Esteban82 commented:
No, no. Just a few. Don't worry.

Esteban82 commented:
The maps of the oceans/seas 

[Seas.pdf](https://github.com/GenericMappingTools/gmt/files/7653263/Seas.pdf)


Data file: I got an error when I try to add the UN countries regions. 
[dcw-collections.txt](https://github.com/GenericMappingTools/gmt/files/7653269/dcw-collections.txt)

anbj commented:
Great work, @Esteban82!

Two corrections (from `HRmLOS_1.1.shp` I guess):

* Barentsz Sea -> Barents Sea
* Skaggerak -> Skagerrak

PaulWessel commented:
Great @Esteban82, can you make a PR for dcw-gmt? That way we can help with fixing these and other spelling issues form the shape file.

PaulWessel commented:
> Data file: I got an error when I try to add the UN countries regions.

What kind of error?

Esteban82 commented:
> I am getting this error when I add an extra record in the dcw-collections.txt. 
> 
> ```
> tag: IHO52 Japan Sea
> region: 125.80325137300008/142.25583300000005/32.57838602500004/51.73925752800005
> 
> gmt coast -Bagf -Ggray -png mapa -N1 -JW15c -RIHO55
> double free or corruption (out)
> Abortado (`core' generado)
> 
> ```
> I suspect that the is somehow related to the number of records because the records that I add it is the number 101. 
> 

This error @PaulWessel

PaulWessel commented:
OK, will look.

PaulWessel commented:
I placed your daw-collections.txt in where DCW lives and it ran fine.  Just to check, the file you posted above is the one that crashes? It has IHO52 on record 172.

PaulWessel commented:
OK, was able to reproduce and then fix it.

Esteban82 commented:
Great. I was a bit confused. I check that the file was ok. When I run an script with `batch ` to make the pdf I have no problem. But I got an error with

`gmt coast -RIHO1 -JQ10c -Gdarkgray -Slightblue -B -png mapa -N1 -Cdarkgray`

Esteban82 commented:
> OK, I think you should define the collections (Med W + Med E) using list: IHO28A, IHO28B.
> 
> Agree on sticking to the original and A or a. I made the checks case insensitive but the _dcw-collections.txt_ file should stick to one syntax (I think I prefer the upper case for tags).

I am not sure if I understand correctly this. What do you mean with should stick to one syntax?

Because, in the codes of the IHO seas, I used upper or lower case following the official documentation. In some cases the used an upper case letter (e.g. IHO15A	Labrador Sea) and in other lower case (e.g. IHO1a Gulf of Bothnia). 

But the problem is that decided is the Mediterranean Sea where they used 28A for the Western Basin and 28a for the Strait of Gibraltar. 

Now, in the dcw-collections I used 28-1 for Med W (but it is not the official tag). So I think that the tag should be tag case sensitive (at least for the IHO codes).

PaulWessel commented:
Aarg, so then we are full circle. Some folks wanted things to be case independent because, I guess, it is too hard to hold the shift key.  But if there are IHO28A and IHO28a then we have a problem.  Actually, IHO have a problem in that they made a stupid decision, but not much we can do about that.

Need to think about this a bit.

Esteban82 commented:
I have a doubt. 

Once approved, how will the DCW collection upgrade work? Will there be releases? Will GMT compare the local version with the server version and update automatically? If the latter, we could add new regions/lists at any time.

PaulWessel commented:
Most GMT users probably install GMT as well as GSHHG and DCW either via the install tarballs or they get them from their Linux distribution.  They will need to reinstall a newer version manually.  Last year or 2019 we added the ability for GMT to download the latest version from the gmt server if these data were not already on the users computer, and it them places those files in ~/.gmt/geography.  However, there is no refresh mechanism for this such as there is for the remote grids and the cache files (where we will detect a new version on the server).  We can discuss if treating both GSHHG and DCW like other remote data and update on demand, but that will require some coding.  Also, this will not help those who installed via tarballs (unless we change coding) and it is then a bit odd to list DCW and GSHHG installers when in fact you don't really have to do this.

Happy to entertain comments on this as it may impact how we deal with GSHHG 3 when that time comes.

Esteban82 commented:
Ok, so the short answer is not. And we should think that the DWC-Collections.txt as a release that the users won't easily update. 

About GSHHG3 it would be good to be able to automatically update (but it is easy to me say since I won't be doing the coding). But also, I saw that GSHHG 2.3.7 is more than 4 year ago. So I don't think that that database is going to have major modifications with time (like the remote grids that are improve every year). So I think you shouldn't spend time on an automatic way to update the data.

PaulWessel commented:
True, it will take us maybe 1-2 years to actually do a GSHHG3 based on OpenSTreetMap so that one is pretty stable until then.  I think the key issues is that we currently have this system

1. GSHHG (and DCW) can be installed from a tar ball.  It will not self-update of course
2. GSHHG (and DCW) can be installed from most Linux bistro.  These will update when we make releases.
3. GSHHG (and DCW), if not installed, will automatically be fetched from the GMT server.  Currently there are no refresh after initial download.

As long as we plan to distribute the official releases, we should probably consider these modifications to these 3 paths so that everybody can stay current or be told to update. Below, each number goes with the methods above:

1. Have GMT check from time to time if there is a new version on the server and notify the user with a static setup.
2. These are self-regulating
3. Add a refreshing scheme

Seems reasonable?

joa-quim commented:
4. GSHHG (and DCW), are automatically installed by the Windows installer and we **never** had an issue with the coastlines on Windows.

That's how we used to do many years ago for all OSs until it was decoupled from GMT (I never liked that move). Installing the coastlines has always been a major pain point when installing on *nix. I hate when someone ask me help to install GMT because I know that I'll dive in docs to relearn how to do it.

I think we should go back to (re)install them at every GMT release and possibly leave the future GSHHG3 as an optional download via some new option in a GMT module because it risks to become a somewhat big file.

PaulWessel commented:
> 7\. GSHHG (and DCW), are automatically installed by the Windows installer and we **never** had an issue with the coastlines on Windows.
> 
> That's how we used to do many years ago for all OSs until it was decoupled from GMT (I never liked that move). Installing the coastlines has always been a major pain point when installing on *nix. I hate when someone ask me help to install GMT because I know that I'll dive in docs to relearn how to do it.
> 
> I think we should go back to (re)install them at every GMT release and possibly leave the future GSHHG3 as an optional download via some new option in a GMT module because it risks to become a somewhat big file.

We also did this because Win users (and now many macOS users) are not that capable of compiling their own GMT version, so an all-including installer (like the macOS bundle) makes sense.  The Linux distributions handle this via packages and dependencies, so I wonder who is building from source.  Perhaps just those needing master and using git and have the skillsets to do this regardless.

And for most of the years, GMT releases are more often than DCW/GSHHG releases so the next installer wipes and installs the latest version for Win and macOS.

anbj commented:
My `dcw-collections.txt`:

```
tag: SCA Scandinavia
list: NO,SE,DK

tag: NSCA Not Scandinavia
list: FR,GR,DE

tag: IHO7 Barentsz Sea
region: 16.581500000000062/68.55576621200004/66.81647200000003/81.82700060000008
```
List everything in `dcw-collections.txt`:

```
$ gmt pscoast -E+n
# TAG	Name		Codes|Region
SCA	Scandinavia	NO,SE,DK
NSCA	Not Scandinavia	FR,GR,DE
IHO7	Barentsz Sea	16.581500000000062/68.55576621200004/66.81647200000003/81.82700060000008
```

Show tags containing `NO` will also return all entries made up of `w/e/s/n (which is a long list when including e.g. the IHO collection):

```
$ gmt pscoast -ENO+n
# TAG	Name		Codes|Region
SCA	Scandinavia	NO,SE,DK
IHO7	Barentsz Sea	16.581500000000062/68.55576621200004/66.81647200000003/81.82700060000008

PaulWessel commented:
Error fixed, @Esteban82

Esteban82 commented:
> Error fixed, @Esteban82

Thanks. Now there is the issue that anbj reported. With `gmt pscoast -ENO+n` I get the collections which include NO and all the regions define by a bounding box (seas, oceans, continents, islands, etc.).

Another issue is that the list follow the original order of the dcw-collections.txt which is not alphabetically and can be confusing for the user.

Esteban82 commented:
If a run `gmt coast -REurope -W -B -png map` I get the map that correspond to `-R=EU`. I think that I should get the bounding box region define in `dcw-collections.txt`. The same happened with Africa, Asia, North America and South America. 

BTW, I noticed that some regions have the same name that collection of countries. 

![map](https://user-images.githubusercontent.com/17908951/149260069-856c1a97-1996-4148-a811-925cdd718361.png)

PaulWessel commented:
Try this update.

Esteban82 commented:
> Try this update.

It works. Now I have to update dcw-collections.txt

PaulWessel commented:
OK, so I can merge this branch then?

Esteban82 commented:
> OK, so I can merge this branch then?

I think that anbj [comment](https://github.com/GenericMappingTools/gmt/pull/6064#issuecomment-989622026) wasn't fix. 

Shortly, If I run `gmt pscoast -ENO+n` I should get only the collections with Norway. But I get also all the bounding box regions.

PaulWessel commented:
Ok, after dinner

PaulWessel commented:
I think I have fixed @anbj's problem, perhaps @Esteban82 can check - I am off for a few hours on an errand.

Esteban82 commented:
> I think I have fixed @anbj's problem, perhaps @Esteban82 can check - I am off for a few hours on an errand.

Great. Now it is fixed.

So the code basically search the string in Codes and in names, right? 

BTW I just noticed that I can make this which is good. 

```
gmt pscoast -EIsland+n
# TAG	Name		Codes|Region
NGNI	New Guinea Island	130.93/150.877/-10.705/-0.339
BRNI	Borneo Island	108.845/119.275/-4.185/7.035
BFFI	Baffin Island	-89.543/-61.263/61.863/73.737
SMTI	Sumatra Island	95.195/106.087/-5.945/5.661
HNSI	Honshu Island	130.862/142.07/33.435/41.546
VCTI	Victoria Island	-119.123/-100.846/68.461/73.378
GBRI	Great Britain Island	-6.231/1.766/49.957/58.673
ELLI	Ellesmere Island	-91.912/-61.101/76.131/83.213
SLWI	Sulawesi Island	118.753/125.252/-5.707/1.756
SNZI	South Island	166.425/174.319/-46.679/-40.499
JAVI	Java Island	105.206/114.603/-8.782/-5.875
NNZI	North Island	172.644/178.553/-41.615/-34.396
...
```

PaulWessel commented:
Given 2.1.0 has been released I think it is OK to approve and merge this branch. Stuff seems to work, e.g.

```
gmt pscoast -E"Inner Seas of the West Coast of Scotland"
-R-8.2827/-4.4275/54.3458/58.6227
```

Any remaining issues, @Esteban82 ?

PaulWessel commented:
Sorry, missed seeing the approval.  Merging

anbj commented:
Nitpicking:

If you define a non-existing code, and at least one existing code, you get a nice warning.

```
$ gmt pscoast -R'i do not exist',NO -JM15c -pdf test -W
coast [WARNING]: No country code matching i do not exist (skipped)
```

If you only define a single non-existing code, you will get a lon/lat- and projection error.
Maybe check if only one input code is given, and if it does not exist, give an (one) error regarding this and stop?

```
$ gmt pscoast -R'i do not exist' -JM15c -pdf test -W
coast [WARNING]: No country code matching i do not exist (skipped)
coast [ERROR]: South is outside -90 to +90 degree range
coast [ERROR]: General map projection error
```
